src/libcore/array/mod.rs

```diff
@@ -387,6 +387,76 @@ where
     }
 }
 
+/// Non-zero numbers
+///
+/// We require following from this trait:
+///  - [(); 0] does **not** implement NonZero.
+///  - [(); N] where N != 0 **implements** NonZero.
+///  As side effect of selected approach, this trait is also implemented
+///  for all other types, but it does not matter.
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+pub auto trait NonZero {}
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+impl !NonZero for [(); 0] {}
+
+// This is implementation for all non-zero-sized arrays
+#[stable(since = "1.4.0", feature = "array_default")]
+impl<T: Default, const N: usize> Default for [T; N]
+where
+    [(); N]: NonZero,
+    [T; N]: LengthAtMost32,
+{
+    #[inline]
+    fn default() -> [T; N] {
+        use crate::mem::MaybeUninit;
+        let mut data: MaybeUninit<[T; N]> = MaybeUninit::uninit();
```

Comment:
Why not make the guard type itself hold the MaybeUninit?

Reply:
Previous version did so.
I tried keeping MaybeUninit outside of Guard to see whether it simplifies code and whether it optimizes generated ASM.